packages/contracts/contracts/protocol/incentives/ExternalRewardDistributor.sol

```diff
@@ -106,44 +143,71 @@ contract ExternalRewardDistributor is IExternalRewardsDistributor {
     uint256 amount
   ) internal {
     address underlying = IAToken(msg.sender).UNDERLYING_ASSET_ADDRESS();
-    uint64 trancheId = IAToken(msg.sender)._tranche();
 
     IERC20(underlying).safeTransferFrom(msg.sender, address(this), amount);
-    IStakingRewards(stakingData[underlying][trancheId]).stake(amount);
+    stake(msg.sender, amount);
 
     // event emission necessary for off chain calculation of looping through all active users
-    emit UserDeposited(user, underlying, trancheId, amount);
+    emit UserDeposited(user, msg.sender, amount);
   }
 
   function onWithdraw(
     address user,
     uint256 amount
   ) internal {
     address underlying = IAToken(msg.sender).UNDERLYING_ASSET_ADDRESS();
-    uint64 trancheId = IAToken(msg.sender)._tranche();
 
-    IStakingRewards(stakingData[underlying][trancheId]).withdraw(amount);
+    unstake(msg.sender, amount);
     IERC20(underlying).safeTransfer(msg.sender, amount);
 
     // event emission necessary for off chain calculation of looping through all active users
-    emit UserWithdraw(user, underlying, trancheId, amount);
+    emit UserWithdraw(user, msg.sender, amount);
   }
 
   function onTransfer(address user, uint256 amount, bool sender) internal {
-    address underlying = IAToken(msg.sender).UNDERLYING_ASSET_ADDRESS();
-    uint64 trancheId = IAToken(msg.sender)._tranche();
     //no-op since totalStaked doesn't change, the amounts each person owns is calculated off chain
-    emit UserTransfer(user, underlying, trancheId, amount, sender);
+    emit UserTransfer(user, msg.sender, amount, sender);
   }
 
   function getStakingContract(address aToken) external view returns (address stakingContract) {
-      address underlying = IAToken(aToken).UNDERLYING_ASSET_ADDRESS();
-      uint64 trancheId = IAToken(aToken)._tranche();
-      return stakingData[underlying][trancheId];
+      return stakingData[aToken];
   }
 
-  function harvestReward(address stakingContract) external onlyManager {
-      IStakingRewards(stakingContract).getReward();
+  /**
+     * @dev harvests rewards in a specific staking contract, and emits an event for subgraph to track rewards claiming to distribute to users
+     * anyone can call this function to harvest
+     * @param stakingContract The contract that staked the tokens and collected rewards
+     * @param assetType The type of the asset (determines the abi of the staking contract)
+     * @param rewardTokens List of reward tokens to claim. For yearn, use the yv Vault that was originally staked in the stakingContract
+     **/
+  function harvestReward(address stakingContract, DataTypes.ReserveAssetType assetType, address[] memory rewardTokens) external {
+
+      uint256[] memory earned = new uint256[](rewardTokens.length);
+      if(assetType == DataTypes.ReserveAssetType.VELODROME) {
+        uint256[] memory balanceBefore = new uint256[](rewardTokens.length);
+        for(uint256 i = 0;i<rewardTokens.length;++i) {
```

Comment:
with event having single rewardToken and earned(instead of array), this is simplified